### DISCUSSION TITLE
컴포넌트 기반으로 개발을 하다 보면 “Props drilling” 문제를 만나게 됩니다.

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#65] [프론트 기술 질문 공간]
1. “Props drilling”은 어떤 문제인가요?
    1. “Props drilling” 문제를 어떻게 해결하셨나요?
    2. 왜 그렇게 해결하는 것이 좋다고 생각하셨나요?

**Comment by michoball** [comment]
Props drilling 은 컴포넌트 기반 개발에서 
상위 컴포넌트에서 하위 컴포넌트로 특정 값은 전달하는 과정에서 발생할 수 있습니다. 
일반적으로 컴포넌트 간 데이터는 props를 통해서 전달 됩니다. 
이때 컴포넌트 계층이 한두 단계정도로만 나뉘어져있다면 상관없지만 계층이 깊어질수록 데이터 관리가 어려워진다는 단점이 있습니다. 

1 특정 컴포넌트 계층에서만 props drilling 문제가 발생한다면 context API 를 사용하는 것만으로도 충분하다고 생각합니다. 
하지만 앱 전체에 걸쳐서 전역저으로 사용되는 데이터라면은 redux, recoil 과 같은 전역 상태관리 툴을 사용하는 것이 좋습니다. 

2 일부 컴포넌트에 걸쳐서 있는 문제라면 굳이 다른 라이브러리를 사용해서 의존성을 높일 필요는 없다고 생각합니다. 하지만 전체 컴포넌트에 걸쳐있는 문제라면 전역 상태관리에 좀 더 특화된 redux와 같은 라이브러리를 상요하는 것이 코드의 가독성을 높이고 좋을 것이라 생각합니다.  저는 특히 redux의 많은 보일러플레이트를 줄인 redux-toolkit 를 많이 사용해봐서 redux-toolkit을 사용할 것 같습니다.

**Comment by WooGie911** [comment]
“Props drilling”이란 상위 컴포넌트에서 하위 컴포넌트로 props를 전달하는 과정에서 너무 많은 컴포넌트를 거치게 되면 코드를 읽을 때 해당 props를 추적하기가 힘들어지는 것을 의미합니다.

1. 변수를 전역관리하여 “Props drilling”을 해결한 경험이 있습니다. 그 과정에서 규모가 아주 큰 경우가 아니라면 가능한 context API 를 사용하고 규모가 큰경우 redux나 redux-toolkit과 같은 라이브러리를 사용하여 전역관리를 한 경험이 있습니다.

2. 규모가 작은 경우 context API를 사용하여 전역관리를 간단하게 해결함으로써 “Props drilling”도 해결할수 있지만 규모가 커지면 관리해야할 상태들이 늘어나면 유지보수가 어려워집니다. 이런 경우 redux나 redux-toolkit을 사용하면 보일러 플레이트가 크다는 명확한 단점이 있지만 flux 패턴의 특징 덕분에 코드의 유지보수가 편하다는 장점 때문에 주로 사용했습니다.

**Comment by BaekYuri** [comment]
Props drilling은 1->2->3->4 와 같이, 상위 컴포넌트에서 하위 컴포넌트로 props를 전달해야 할 때 너무 많은 컴포넌트를 거쳐 전달되는 경우를 말합니다.
1. props drilling을 해결하기 위해 Redux를 이용하여 전역 상태 관리를 했던 경험이 있습니다. 여러 컴포넌트에서 어떠한 변수를 사용한다고 생각이 든다면, redux에 저장하여 사용하였고, 간단하게 한 두개의 컴포넌트에서만 사용하는 변수라면 props를 이용해 구현했던 경험이 있습니다.
2. 한 두개의 컴포넌트에서만 사용하는 변수를 redux에 저장해서 사용한다면 오히려 가독성을 해칠 수 있다고 생각하였기 때문에 간단하게 구현하였습니다.

**Comment by javascriptbiggosoo** [comment]
1. Recoil을 사용하면 어느 컴포넌트에서든 state를 꺼내 쓸 수 있기 때문에 여러 계층의 컴포넌트에 props를 내려주지 않아도 됩니다.
2. 성능 최적화 면에서 컴포넌트는 넘겨받는 state가 변경될 때만 다시 렌더링되므로, 여러 계층의 컴포넌트에 props를 내려주는 것보다 불필요한 렌더링을 줄일 수 있습니다.
코드베이스를 단순화하고 이해하고 유지 관리하기 쉽게 만드는 것 또한 무시할 수 없는 덤입니다.
